pkg/controller/oam/applicationconfiguration/apply.go

```diff
@@ -19,61 +19,210 @@ package applicationconfiguration
 	errFmtSetWorkloadRef = "cannot set trait %q reference to %q"
 	errFmtApplyTrait     = "cannot apply trait %q"
+	errFmtApplyScope     = "cannot apply scope %q"
 )
 
 // A WorkloadApplicator creates or updates workloads and their traits.
 type WorkloadApplicator interface {
 	// Apply a workload and its traits.
-	Apply(ctx context.Context, w []Workload, ao ...resource.ApplyOption) error
+	Apply(ctx context.Context, namespace string, status []v1alpha2.WorkloadStatus, w []Workload, ao ...resource.ApplyOption) error
 }
 
 // A WorkloadApplyFn creates or updates workloads and their traits.
-type WorkloadApplyFn func(ctx context.Context, w []Workload, ao ...resource.ApplyOption) error
+type WorkloadApplyFn func(ctx context.Context, namespace string, status []v1alpha2.WorkloadStatus, w []Workload, ao ...resource.ApplyOption) error
 
 // Apply a workload and its traits.
-func (fn WorkloadApplyFn) Apply(ctx context.Context, w []Workload, ao ...resource.ApplyOption) error {
-	return fn(ctx, w, ao...)
+func (fn WorkloadApplyFn) Apply(ctx context.Context, namespace string, status []v1alpha2.WorkloadStatus, w []Workload, ao ...resource.ApplyOption) error {
+	return fn(ctx, namespace, status, w, ao...)
 }
 
 type workloads struct {
-	client resource.Applicator
+	client    resource.Applicator
+	rawClient client.Client
 }
 
-func (a *workloads) Apply(ctx context.Context, w []Workload, ao ...resource.ApplyOption) error {
+func (a *workloads) Apply(ctx context.Context, namespace string, status []v1alpha2.WorkloadStatus, w []Workload, ao ...resource.ApplyOption) error {
 	for _, wl := range w {
 		if err := a.client.Apply(ctx, wl.Workload, ao...); err != nil {
 			return errors.Wrapf(err, errFmtApplyWorkload, wl.Workload.GetName())
 		}
 
+		workloadRef := runtimev1alpha1.TypedReference{
+			APIVersion: wl.Workload.GetAPIVersion(),
+			Kind:       wl.Workload.GetKind(),
+			Name:       wl.Workload.GetName(),
+		}
+
 		for i := range wl.Traits {
 			t := &wl.Traits[i]
 
-			ref := runtimev1alpha1.TypedReference{
-				APIVersion: wl.Workload.GetAPIVersion(),
-				Kind:       wl.Workload.GetKind(),
-				Name:       wl.Workload.GetName(),
-			}
 			// TODO(rz): Need to find a way to make sure that this is compatible with the trait structure.
-			if err := fieldpath.Pave(t.UnstructuredContent()).SetValue("spec.workloadRef", ref); err != nil {
+			if err := fieldpath.Pave(t.UnstructuredContent()).SetValue("spec.workloadRef", workloadRef); err != nil {
 				return errors.Wrapf(err, errFmtSetWorkloadRef, t.GetName(), wl.Workload.GetName())
 			}
 
 			if err := a.client.Apply(ctx, t, ao...); err != nil {
 				return errors.Wrapf(err, errFmtApplyTrait, t.GetName())
 			}
 		}
+
+		for _, s := range wl.Scopes {
+			return a.applyScope(ctx, wl, s)
+		}
+	}
+
+	return a.dereferenceScope(ctx, namespace, status, w)
+}
+
+func (a *workloads) dereferenceScope(ctx context.Context, namespace string, status []v1alpha2.WorkloadStatus, w []Workload) error {
```

Comment:
I don't seem to see any UT for those functions. A good UT will cover the "leaf" functions individually instead of one umbrella test which kinda defies the purpose of a UT.